examples/desktop/events/monitor.c

```diff
@@ -15,113 +16,106 @@
 
-  atclient_atsign atsign;
-  atclient_atsign_init(&atsign, atsign_input);
+  atclient monitor_conn;
+  atclient_monitor_init(&monitor_conn);
 
-  atclient_atkeys atkeys;
-  atclient_atkeys_init(&atkeys);
-  const char *homedir;
+  atclient heartbeat_conn;
+  atclient_monitor_init(&heartbeat_conn);
 
-  if ((homedir = getenv("HOME")) == NULL) {
-    printf("HOME not set\n");
-    return 1;
+  pthread_t tid;
+
+  atclient_monitor_message *message = NULL;
+
+  if ((ret = get_atsign_input(argc, argv, &atsign)) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to get atsign input (Example: \'./monitor -a @bob\')\n");
+    goto exit;
+  }
+
+  if ((ret = set_up_atkeys(&atkeys, atsign, strlen(atsign))) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set up atkeys: %d\n", ret);
+    goto exit;
   }
-  char atkeys_path[1024];
-  snprintf(atkeys_path, 1024, "%s/.atsign/keys/%s_key.atKeys", homedir, atsign_input);
 
-  atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_INFO, "Reading atkeys file: %s\n", atkeys_path);
-  atclient_atkeys_populate_from_path(&atkeys, atkeys_path);
+  if ((ret = atclient_connection_connect(&root_connection, ROOT_HOST, ROOT_PORT)) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to connect to root server\n");
+    goto exit;
+  }
 
-  printf("Starting monitor\n");
-  struct atclient monitor_ctx;
-  atclient_monitor_init(&monitor_ctx, atsign, atkeys);
-  ret = atclient_start_monitor(&monitor_ctx, ROOT_HOST, ROOT_PORT, "");
-  if (ret != 0) {
-    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Monitor crashed\n");
+  if((ret = atclient_monitor_pkam_authenticate(&monitor_conn, &root_connection, &atkeys, atsign)) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to authenticate monitor with PKAM\n");
     goto exit;
   }
-  printf("Monitor started!\n");
 
-  printf("Starting heartbeat\n");
-  pthread_t tid;
-  ret = pthread_create(&tid, NULL, heartbeat_handler, &monitor_ctx);
-  if (ret != 0) {
-    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to create heartbeat_handler\n");
+  if((ret = atclient_monitor_pkam_authenticate(&heartbeat_conn, &root_connection, &atkeys, atsign)) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to authenticate heartbeat with PKAM\n");
     goto exit;
   }
-  printf("Heartbeat started!\n");
-  if (ret < 0) {
+
+  if ((ret = atclient_monitor_start(&monitor_conn, ".*", strlen(".*"))) != 0) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Monitor crashed\n");
     goto exit;
   }
 
-  if ((ret = atclient_pkam_authenticate(&atclient, &root_connection, &atkeys, atsign.atsign)) !=
-      0) {
-    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to authenticate\n");
+  atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_DEBUG, "Starting heartbeat thread\n");
+  if ((ret = pthread_create(&tid, NULL, heartbeat_handler, &heartbeat_conn)) != 0) {
```

Comment:
We don't want to start a separate thread for heartbeat. It is meant to keep the monitor connection alive. If we are having locking problems, we could implement hooks to monitor read & send heartbeat. So that the end developer can add there own semaphores or mutex locks and still maintain full control.

Reply:
We should only have one additional connection for monitor & one additional thread for the heartbeats

Reply:
That thought came across my mind

Reply:
Actually we don't even need hooks, I can just wrap semaphores around each of the function calls.

Reply:
I have no idea why I thought a separate connection would keep the original monitor connection alive